apis/r/R/write_seurat.R

```diff
@@ -207,6 +208,210 @@ write_soma.Assay <- function(
+  uri <- .check_soma_uri(
+    uri = uri,
+    soma_parent = soma_parent,
+    relative = relative
+  )
+
+  # Create the measurement
+  ms <- SOMAMeasurementCreate(
+    uri = uri,
+    ingest_mode = ingest_mode,
+    platform_config = platform_config,
+    tiledbsoma_ctx = tiledbsoma_ctx
+  )
+  ms$set_metadata(.assay_version_hint('v5'))
+  X <- if (!'X' %in% ms$names()) {
+    SOMACollectionCreate(
+      uri = file_path(ms$uri, 'X'),
+      ingest_mode = ingest_mode,
+      platform_config = platform_config,
+      tiledbsoma_ctx = tiledbsoma_ctx
+    )
+  } else if (isTRUE(relative)) {
+    SOMACollectionOpen(uri = file_path(ms$uri, 'X'), mode = 'WRITE')
+  } else {
+    ms$X
+  }
+  withCallingHandlers(
+    .register_soma_object(X, soma_parent = ms, key = 'X', relative = relative),
+    existingKeyWarning = .maybe_muffle
+  )
+  on.exit(X$close(), add = TRUE, after = FALSE)
+
+  X$set_metadata(.layer_hint(SeuratObject::DefaultLayer(x)))
+
+  # Pull presence matrices from the v5 assay
+  cmat <- methods::slot(x, name = 'cells')
+  fmat <- methods::slot(x, name = 'features')
+
+  # Write `X` matrices
+  for (lyr in SeuratObject::Layers(x)) {
+    ldat <- SeuratObject::LayerData(x, layer = lyr)
+    if (!inherits(ldat, what = c("matrix", "Matrix"))) {
+      warning(warningCondition(
+        message = sprintf("%s", lyr),
+        class = "unknownMatrixTypeWarning",
+        call = str2lang("write_soma()")
+      ))
+      next
+    }
+    type <- .type_hint(ifelse(is.matrix(ldat), yes = 'matrix', no = class(ldat)))
+    if (all(fmat[, lyr]) && all(cmat[, lyr])) {
+      spdl::info("Adding '{}' matrix as '{}'", lyr, lyr)
+      tryCatch(
+        expr = {
+          arr <- write_soma(
+            x = ldat,
+            uri = lyr,
+            soma_parent = X,
+            sparse = TRUE,
+            transpose = TRUE,
+            ingest_mode = ingest_mode,
+            shape = shape,
+            key = lyr,
+            platform_config = platform_config,
+            tiledbsoma_ctx = tiledbsoma_ctx
+          )
+          arr$set_metadata(type)
+        },
+        error = function(err) {
+          if (lyr %in% SeuratObject::DefaultLayer(x)) {
+            stop(err)
+          }
+          err_to_warn(err)
+        }
+      )
+      next
+    }
+    ldat <- Matrix::t(as(ldat, "TsparseMatrix"))
+    idx <- which(cmat[, lyr])
+    jdx <- which(fmat[, lyr])
+    coo <- data.frame(
+      soma_dim_0 = bit64::as.integer64(idx[ldat@i + 1L] - 1L),
+      soma_dim_1 = bit64::as.integer64(jdx[ldat@j + 1L] - 1L),
+      soma_data = ldat@x
+    )
+    shape <- c(max(coo$soma_dim_0), max(coo$soma_dim_1)) + 1L
+    arr <- X$add_new_sparse_ndarray(
+      key = lyr,
+      type = arrow::infer_type(coo$soma_data),
+      shape = as.integer(shape)
+    )
+    arr$.__enclos_env__$private$.write_coo_dataframe(coo)
+    arr$set_metadata(.ragged_array_hint())
+    arr$set_metadata(type)
+  }
+
+  # Write feature-level meta data
```

Comment:
```suggestion
  # Write feature-level metadata
```

Reply:
"meta data" is how Seurat spell this out https://satijalab.github.io/seurat-object/reference/sub-sub-.Assay5.html, and was used previously in the v3 -> SOMA ingestor https://github.com/single-cell-data/TileDB-SOMA/blob/main/apis/r/R/write_seurat.R#L172